doc/flow-definition.md

```diff
@@ -1,11 +1,308 @@
+## Naming
 
-* The flow file should be 'example-smart-answer.rb'
-* The flow class should be `ExampleSmartAnswerFlow`
-* The flow name should be 'example-smart-answer'
+The flow filename should be based on the path where the Smart Answer is to be found on gov.uk. For example, for a Smart Answer at https://www.gov.uk/example-smart-answer, the flow file should be named `lib/smart_answer_flows/example-smart-answer.rb`. Similarly the template directory should be named `lib/smart_answer_flows/example-smart-answer/`.
 
-The `Flow` class contains a single `#define` method that defines the questions (see "Question types" below), rules (see "Defining next node rules" below) and outcomes (see "Outcome templates" below).
+The flow class should be a camel-case version of the flow filename with the suffix, `Flow` e.g. in the case above it would be `ExampleSmartAnswerFlow`. The class should inherit from `SmartAnswer::Flow` and be in the `SmartAnswer` namespace.
+
+For example:
+
+```ruby
+# lib/smart_answer_flows/example-smart-answer/example-smart-answer.rb
+module SmartAnswer
+  class ExampleSmartAnswerFlow < Flow
+    def define
+      # flow definition specified here
+    end
+  end
+end
+```
+
+> Note that much of the above does not follow standard Ruby or Rails conventions which isn't ideal. However, the plan is to move towards using such conventions as soon as possible.
+
+## Definition
+
+The flow is defined within the `define` instance method of the `Flow` subclass using a DSL to specify metadata, question nodes and outcome nodes.
+
+> Unfortunately this DSL is overly complex and makes it all too easy to create code which is very hard to follow. The DSL makes heavy use of Ruby's [`BasicObject#instance_eval`][instance-eval] & [`#instance_exec`][instance-exec]. This means that:
+>
+> * It's not obvious what variables are in scope within a given block of code.
+> * It's all too easy to write arbitrary custom code which can make each flow a bit of a special case. This doesn't help on the maintainability front.
+> * It's confusing that blocks of code are not necessarily executed in the order that they appear in the flow definition.
+> * The code in the blocks is not easily unit-testable.
+
+Each flow is instantiated in the `FlowRegistry` via the `Flow.build` method which, in turn, instantiates the flow and invokes its `Flow#define` method.
+
+Currently in production a single instance of each flow class is instantiated at application start-up (or strictly speaking on the first request) and cached in the `FlowRegistry` i.e. the same instance of a particular flow class is used to service all the user requests involving that Smart Answer. Thus it's important that no request-specific state is stored on the flow instance, otherwise this could leak into other user requests.
+
```

Comment:
> What could a request-specific state look like?

When I say "state", I mean the state (in the [generic sense][1]) of any object (e.g. instance variables). When I say "request-specific state", I mean _any_ state that is specific to an individual user request.

It is actually pretty hard, but not impossible, to access the flow instance from within the blocks which are processed at request time. I just wanted to make it clear that the _same_ flow instance is used for _every_ request and so you need to be careful that information related to the request from one user does not leak into the request from another user via the flow instance.

> Is this related to the save input as changes in #2608?

I assume you mean the plan to use `on_response` in preference to `save_input_as`. If so, then no, this paragraph is not related to that.

I've tried, but failed to improve this paragraph. Since there's not anything _incorrect_ about the paragraph as it stands, I suggest that we merge it as it stands and then chat about it as a group. Then if we think it's necessary, we can make improvements.

[1]: https://en.wikipedia.org/wiki/State_(computer_science)